google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/application/AppEngineApplicationCreateDialog.java

```diff
@@ -0,0 +1,206 @@
+        "<a href=\"" + LOCATIONS_DOCUMENTATION_URL + "\">", "</a>") + "</p>" + HTML_CLOSE_TAG);
+  }
+
+  @Override
+  protected void doOKAction() {
+    final Location selectedLocation
+        = ((AppEngineLocationSelectorItem) regionComboBox.getSelectedItem()).getLocation();
+
+    // show loading state
+    setOKActionEnabled(false);
+
+    try {
+      UsageTrackerProvider.getInstance()
+          .trackEvent(GctTracking.APP_ENGINE_APPLICATION_CREATE)
+          .ping();
+
+      // attempt to create the application, and close the dialog if successful
+      ProgressManager.getInstance().runProcessWithProgressSynchronously(() ->
+              AppEngineAdminService.getInstance().createApplication(
+                  selectedLocation.getLocationId(), gcpProjectId, userCredential),
+      GctBundle.message("appengine.application.create.loading",
+              selectedLocation.getLocationId()),
+          true /* cancellable */,
+          ProjectManager.getInstance().getDefaultProject());
+
+      UsageTrackerProvider.getInstance()
+          .trackEvent(GctTracking.APP_ENGINE_APPLICATION_CREATE_SUCCESS)
+          .ping();
+
+      close(OK_EXIT_CODE);
+
+    } catch (IOException e) {
+      trackApplicationCreateFailure();
+      setStatusMessage(GctBundle.message("appengine.application.create.error.transient"), true);
+      setOKActionEnabled(true);
+      return;
+    } catch (GoogleApiException e) {
+      trackApplicationCreateFailure();
+      setStatusMessage(e.getMessage(), true);
+      setOKActionEnabled(true);
+      return;
+    } catch (Exception e) {
+      trackApplicationCreateFailure();
+      throw new RuntimeException(e);
+    }
+  }
+
+  private void trackApplicationCreateFailure() {
+    UsageTrackerProvider.getInstance()
+        .trackEvent(GctTracking.APP_ENGINE_APPLICATION_CREATE_FAIL)
+        .ping();
+  }
+
+  private void setStatusMessageAsync(final String message, final boolean isError) {
+    ApplicationManager.getApplication().invokeLater(() -> setStatusMessage(message, isError),
+        ModalityState.stateForComponent(AppEngineApplicationCreateDialog.this.getContentPane()));
+  }
+
+  private void setStatusMessage(String message, boolean isError) {
+    statusPane.setText(message);
+    statusPane.setForeground(isError ? JBColor.red : JBColor.black);
+    statusPane.setVisible(true);
+  }
+
+  private void refreshLocationsSelector() {
+    ApplicationManager.getApplication().executeOnPooledThread(() -> {
+      final List<Location> appEngineRegions;
+      try {
+        appEngineRegions = AppEngineAdminService.getInstance()
+            .getAllAppEngineLocations(userCredential);
+      } catch (IOException | GoogleApiException e) {
+        setStatusMessageAsync(GctBundle.message("appengine.application.region.list.fetch.error"),
+            true);
+        return;
+      }
+
+      // perform the actual UI updates on the event dispatch thread
+      ApplicationManager.getApplication().invokeLater(() -> {
```

Comment:
UI updates need to be performed from the swing event dispatch thread. Since we want to update the list of regions from a background thread, we need to use `Application.invokeLater` in order to execute the code that modifies the UI within the event dispatch thread.